pkg/client/cache/thread_safe_store.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// ThreadSafeStore is an interface that allows concurrent access to a storage backend.
+// Note that the guarantees of thread safety provided by List/Get are only valid
+// if the caller treats returned items as read-only. For example, a pointer inserted
+// in the store through `Add` will be returned as is by `Get`. Multiple clients might
+// invoke `Get` on the same key and modify the pointer in a non-thread-safe way.
```

Comment:
Hmm... now that I look again, maybe this warning is insufficient. Maybe it's better to say "you must not modify anything returned by get or list, it would totally break the indexing feature in addition to being not thread safe".